voice/src/main/kotlin/gateway/VoiceEvent.kt

```diff
@@ -3,71 +3,68 @@ package dev.kord.voice.gateway
-                            OpCode.SessionDescription -> decodeSerializableElement(
-                                descriptor,
-                                index,
-                                SessionDescription.serializer()
-                            )
-                            OpCode.Speaking -> decodeSerializableElement(descriptor, index, Speaking.serializer())
-                            OpCode.Resumed -> Resumed
-                            else -> {
-                                val element = decodeNullableSerializableElement(
-                                    descriptor,
-                                    index,
-                                    JsonElement.serializer().nullable
-                                )
-
-                                jsonLogger.debug { "Unknown event with Opcode $op : $element" }
-                                null
-                            }
-                        }
-                    }
+            var d: JsonElement? = null
+            while (true) {
+                when (val index = decodeElementIndex(descriptor)) {
+                    0 -> op = decodeSerializableElement(descriptor, index, OpCode.serializer(), op)
+                    1 -> d = decodeSerializableElement(descriptor, index, JsonElement.serializer(), d)
+                    CompositeDecoder.DECODE_DONE -> break
+                    else -> throw SerializationException("Unexpected index: $index")
+                }
+            }
+            when (op) {
+                null ->
+                    throw @OptIn(ExperimentalSerializationApi::class) MissingFieldException("op", descriptor.serialName)
+                OpCode.Ready -> decodeEvent(decoder, op, Ready.serializer(), d)
+                OpCode.SessionDescription -> decodeEvent(decoder, op, SessionDescription.serializer(), d)
+                OpCode.Speaking -> decodeEvent(decoder, op, Speaking.serializer(), d)
+                OpCode.HeartbeatAck -> decodeEvent(decoder, op, HeartbeatAck.serializer(), d)
+                OpCode.Hello -> decodeEvent(decoder, op, Hello.serializer(), d)
+                OpCode.Resumed -> {
+                    // ignore the d field, Resumed is supposed to have null here:
+                    // https://discord.com/developers/docs/topics/voice-connections#resuming-voice-connection-example-resumed-payload
+                    Resumed
                 }
-                endStructure(descriptor)
-                return data
+                OpCode.ClientDisconnect -> null // TODO what should we do with this event?
+                // OpCodes for Commands, they shouldn't be received
+                OpCode.Identify, OpCode.SelectProtocol, OpCode.Heartbeat, OpCode.Resume ->
+                    throw IllegalArgumentException("Illegal opcode for voice gateway event: $op")
+                OpCode.Unknown -> null // TODO should we throw an exception instead?
```

Comment:
@DRSchlaubi @lost-illusi0n do you know what we should do for the [Client Disconnect](https://discord.com/developers/docs/topics/opcodes-and-status-codes#voice-voice-opcodes) opcode? It's not mentioned in the [voice documentation](https://discord.com/developers/docs/topics/voice-connections).

Also how likely is it to encounter unknown opcodes? If it isn't expected to happen, should we throw an exception for unknown opcodes like we do for the [non-voice gateway](https://github.com/kordlib/kord/blob/d1651a756f8db73909b0aa489040d63f32943619/gateway/src/commonMain/kotlin/Event.kt#L73)?

Reply:
A lot of the voice protocol isn't documented (see my prs lol)

The Client Disconnect op is sent every time a user disconnects so it's probably best to ignore it to prevent spam

Reply:
alright, i should probably add back the debug logging for this and then it's ok for now

Reply:
> A lot of the voice protocol isn't documented (see my prs lol)

So ig that means the answer to how likely it is to encounter unknown op codes is VERY LIKELY